login: fix redirect loop on formkey or nonce mismatch

justcool393 opened the pull request:
needs testing but should be good regardless

justcool393 committed:
fix 401-302-401-302-429 loop

justcool393 committed:
don't logout users on bad form key, just treat the request as unauthenticated

TLSM reviewed (approved):
This identical to logic on the upstream that's been working in prod for at least a few days, and nearly identical to logic that's been working for a number of weeks.

zorbathut commented:
Looks like the `test_signup_and_post` test doesn't like this; maybe changes to the login path caused it to break? It may be overly fragile but I still want it fixed :)

TLSM committed:
Handle None/empty case in validate_formkey.

A supplied empty formkey, or the lack of a supplied formkey (None) is not a valid formkey. Handle this inside the function rather than at the call-site.

TLSM commented:
Looks to have been a legitimate test failure, not a false positive! Good catch by the test. Fixed now.

…Turned out we moved the None-check for `validate_formkey` into the function itself on upstream, which did not get ported over with the rest of the changes.

justcool393 committed:
Validate as false if no hashstr or string
